include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+
+namespace rc {
+namespace state {
+namespace gen {
+
+namespace detail {
+template <typename Cmd, typename GenFunc>
+class ParallelCommandsGen {
+public:
+  using CmdSP = std::shared_ptr<const Cmd>;
+  using Model = typename Cmd::Model;
+  using Sut = typename Cmd::Sut;
+
+  template <typename ModelArg, typename GenFuncArg>
+  ParallelCommandsGen(ModelArg &&initialState, GenFuncArg &&genFunc)
+      : m_initialState(std::forward<ModelArg>(initialState))
+      , m_genFunc(std::forward<GenFuncArg>(genFunc)) {}
+
+  Shrinkable<ParallelCommands<Cmd>> operator()(const Random &random,
+                                               int size) const {
+    return generateCommands(random, size);
+  }
+
+private:
+  struct CommandEntry {
+    CommandEntry(Random &&aRandom, Shrinkable<CmdSP> &&aShrinkable)
+        : random(std::move(aRandom))
+        , shrinkable(std::move(aShrinkable)) {}
+
+    Random random;
+    Shrinkable<CmdSP> shrinkable;
+  };
+
+  struct CommandSequence {
+    CommandSequence(const GenFunc &func, int sz)
+        : genFunc(func)
+        , size(sz) {}
+
+    Model initialState;
+    GenFunc genFunc;
+    int size;
+    std::vector<CommandEntry> entries;
+
+
+    void repairEntriesFrom(std::size_t start) {
+      for (auto i = start; i < entries.size(); i++) {
+        if (!repairEntryAt(i)) {
+          entries.erase(begin(entries) + i--);
+        }
+      }
+    }
+
+    bool repairEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        const auto cmd = entry.shrinkable.value();
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+
+        return regenerateEntryAt(i);
+      }
+
+      return true;
+    }
+
+    bool regenerateEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        entry.shrinkable = genFunc()(entry.random, size);
+        const auto cmd = entry.shrinkable.value();
+        return true;
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+      } catch (const GenerationFailure &failure) {
+        // Just return false below
+      }
+
+      return false;
+    }
+  };
+
+  struct ParallelCommandSequence {
+    ParallelCommandSequence(CommandSequence prefix,
+                            CommandSequence left,
+                            CommandSequence right)
+        : prefix(prefix)
+        , left(left)
+        , right(right) {}
+
+    CommandSequence prefix;
+    CommandSequence left;
+    CommandSequence right;
+  };
+
+  Shrinkable<ParallelCommands<Cmd>> generateCommands(const Random &random,
+                                                     int size) const {
+    return shrinkable::map(generateParallelSequence(random, size),
+                           [](const ParallelCommandSequence &sequence) {
```

Comment:
All of the `::map` combinators pass their arguments as rvalues so you could probably declare this as `ParallelCommandSequence &&sequence` if you wanted. However, there's no point here since you're not actually copying/moving the shrinkables, you´re moving their `value()`s which will be rvalues regardless.